[lldb] [mostly NFC] Large WP foundation: WatchpointResources

jasonmolenda opened the pull request:
This patch is rearranging code a bit to add WatchpointResources to Process.  A WatchpointResource is meant to represent a hardware watchpoint register in the inferior process.  It has an address, a size, a type, and a list of Watchpoints that are using this WatchpointResource.

This current patch doesn't add any of the features of WatchpointResources that make them interesting -- a user asking to watch a 24 byte object could watch this with three 8 byte WatchpointResources.  Or a Watchpoint on 1 byte at 0x1002 and a second watchpoint on 1 byte at 0x1003, these must both be served by a single WatchpointResource on that doubleword at 0x1000 on a 64-bit target, if two hardware watchpoint registers were used to track these separately, one of them may not be hit.  Or if you have one Watchpoint on a variable with a condition set, and another Watchpoint on that same variable with a command defined or different condition, or ignorecount, both of those Watchpoints need to evaluate their criteria/commands when their WatchpointResource has been hit.

There's a bit of code movement to rearrange things in the direction I'll need for implementing this feature, so I want to start with reviewing & landing this mostly NFC patch and we can focus on the algorithmic choices about how WatchpointResources are shared and handled as they're triggeed, separately.

This patch also stops printing "Watchpoint <n> hit: old value: <x>, new vlaue: <y>" for Read watchpoints.  I could make an argument for print "Watchpoint <n> hit: current value <x>" but the current output doesn't make any sense, and the user can print the value if they are particularly interested.  Read watchpoints are used primarily to understand what code is reading a variable.

This patch adds more fallbacks for how to print the objects being watched if we have types, instead of assuming they are all integral values, so a struct will print its elements.  As large watchpoints are added, we'll be doing a lot more of those.

To track the WatchpointSP in the WatchpointResources, I changed the internal API which took a WatchpointSP and devolved it to a Watchpoint*, which meant touching several different Process files. I removed the watchpoint code in ProcessKDP which only reported that watchpoints aren't supported, the base class does that already.

I haven't yet changed how we receive a watchpoint to identify the WatchpointResource responsible for the trigger, and identify all Watchpoints that are using this Resource to evaluate their conditions etc.  This is the same work that a BreakpointSite needs to do when it has been tiggered, where multiple Breakpoints may be at the same address.

There is not yet any printing of the Resources that a Watchpoint is implemented in terms of ("watchpoint list", or
SBWatchpoint::GetDescription).

"watchpoint set var" and "watchpoint set expression" take a size argument which was previously 1, 2, 4, or 8 (an enum).  I've changed this to an unsigned int.  Most hardware implementations can only watch 1, 2, 4, 8 byte ranges, but with Resources we'll allow a user to ask for different sized watchpoints and set them in hardware-expressble terms soon.

I've annotated areas where I know there is work still needed with LWP_TODO that I'll be working on once this is landed.

I've tested this on aarch64 macOS, aarch64 Linux, and Intel macOS.

https://discourse.llvm.org/t/rfc-large-watchpoint-support-in-lldb/72116

jasonmolenda committed:
[lldb] [mostly NFC] Large WP foundation: WatchpointResources

This patch is rearranging code a bit to add WatchpointResources to
Process.  A WatchpointResource is meant to represent a hardware
watchpoint register in the inferior process.  It has an address, a
size, a type, and a list of Watchpoints that are using this
WatchpointResource.

This current patch doesn't add any of the features of WatchpointResources
that make them interesting -- a user asking to watch a 24 byte
object could watch this with three 8 byte WatchpointResources.  Or
a Watchpoint on 1 byte at 0x1002 and a second watchpoint on 1 byte
at 0x1003, these must both be served by a single WatchpointResource
on that doubleword at 0x1000 on a 64-bit target, if two hardware
watchpoint registers were used to track these separately, one of
them may not be hit.  Or if you have one Watchpoint on a variable
with a condition set, and another Watchpoint on that same variable
with a command defined or different condition, or ignorecount, both
of those Watchpoints need to evaluate their criteria/commands when
their WatchpointResource has been hit.

There's a bit of code movement to rearrange things in the direction
I'll need for implementing this feature, so I want to start with
reviewing & landing this mostly NFC patch and we can focus on the
algorithmic choices about how WatchpointResources are shared and
handled as they're triggeed, separately.

This patch also stops printing "Watchpoint <n> hit: old value: <x>,
new vlaue: <y>" for Read watchpoints.  I could make an argument for
print "Watchpoint <n> hit: current value <x>" but the current output
doesn't make any sense, and the user can print the value if they
are particularly interested.  Read watchpoints are used primarily
to understand what code is reading a variable.

This patch adds more fallbacks for how to print the objects being
watched if we have types, instead of assuming they are all integral
values, so a struct will print its elements.  As large watchpoints
are added, we'll be doing a lot more of those.

To track the WatchpointSP in the WatchpointResources, I changed
the internal API which took a WatchpointSP and devolved it to a
Watchpoint*, which meant touching several different Process files.
I removed the watchpoint code in ProcessKDP which only reported
that watchpoints aren't supported, the base class does that already.

I haven't yet changed how we receive a watchpoint to identify
the WatchpointResource responsible for the trigger, and identify
all Watchpoints that are using this Resource to evaluate their
conditions etc.  This is the same work that a BreakpointSite needs
to do when it has been tiggered, where multiple Breakpoints may be
at the same address.

There is not yet any printing of the Resources that a Watchpoint
is implemented in terms of ("watchpoint list", or
SBWatchpoint::GetDescription).

"watchpoint set var" and "watchpoint set expression" take a size
argument which was previously 1, 2, 4, or 8 (an enum).  I've
changed this to an unsigned int.  Most hardware implementations
can only watch 1, 2, 4, 8 byte ranges, but with Resources we'll
allow a user to ask for different sized watchpoints and set
them in hardware-expressble terms soon.

I've annotated areas where I know there is work still needed with
LWP_TODO that I'll be working on once this is landed.

I've tested this on aarch64 macOS, aarch64 Linux, and Intel macOS.

https://discourse.llvm.org/t/rfc-large-watchpoint-support-in-lldb/72116

jasonmolenda commented:
I started on the next step of updating ProcessGDBRemote to identify the WatchpointResource that was hit by a watchpoint stop, and evaluate all of its Watchpoints, when I looked at BreakpointSite and BreakpointSiteList and realize I should have started by coping their design/functionality.  I'm rewriting WatchpointResource / WatchpointResourceList right now.

DavidSpickett commented on lldb/test/Shell/Watchpoint/LocalVariableWatchpointDisabler.test:
```diff
@@ -1,3 +1,10 @@
+# but the lines being output by lldb are identical,
+# by visual inspection.  
+# FileCheck is seeing some difference between them,
+# which I need to get to the bottom of.
```
A trick I used to do for python doctests was `str.replace(" ", "?")`. Then you could see the trailing spaces. Might work here if you have that or extra newlines.

DavidSpickett commented on lldb/test/API/python_api/watchpoint/watchlocation/TestTargetWatchAddress.py:
```diff
@@ -135,5 +135,5 @@ def test_watch_address_with_invalid_watch_size(self):
                 error.GetCString(),
                 exe=False,
-                substrs=["watch size of %d is not supported" % 365],
+                substrs=["Setting one of the watchpoint resources failed"],
```
Is the original error still relevant? Should it be one of the watchpoint resources failed because...

jasonmolenda replied:
This test is trying to watch a 365 byte region, it was intended to detect the fact that you can only watch 1/2/4/8 byte blocks of memory (the --size was really an enum and it had to be one of those).  With this patch currently, lldb will accept the 365 byte watch request and pass it to the remote stub, which will probably return an error.  (debugserver will actually break it into aligned watchable regions and do the right thing). When I finish the large watchpoint work, lldb will break this 365 byte request into 46 8-byte regions (for linux with BAS only watchpoints) and when we try to set the 5th one (or however many the core actually supports), that watch request will fail.  Maybe the error message should say something about how your watchpoint was broken into <x> watchpoint requests but only <y> of them could be set.

DavidSpickett replied:
Yeah I don't think it would need to be super specific, that is what logs are for. Drawing the distinction between failed because the stub didn't have resources, and failed because the stub exploded, that's more what I'd like to see.

And here the original error message is actually still pretty vague. It tells me that size N isn't support, but not what is supported (which is a whole other thing).

DavidSpickett commented on lldb/test/API/commands/watchpoints/watchpoint_count/main.c:
```diff
@@ -2,8 +2,8 @@
-  uint8_t x1 = 0;
-  uint16_t x2 = 0;
+  long x1 = 0;
+  long x2 = 0;
```
We need the type change here because?

DavidSpickett commented on lldb/source/Target/WatchpointResourceList.cpp:
```diff
@@ -0,0 +1,61 @@
+  if (idx < m_resources.size()) {
+    return m_resources[idx];
+  } else {
+    return {};
```
Single line ifs.

DavidSpickett replied:
Also return in an else (as in, no need for the else)

jasonmolenda committed:
Restructure WatchpointResources to be closer to BreakpointSites

Restructure the way I added WatchpointResources to be
much more in line with how BreakpointLocations relate to
BreakpointSites.  Similar to BreakpointSites, when we have
a watchpoint access, we will identify the WatchpointResource
that tiggered it, and evaluate all Watchpoints who own that
WatchpointResource.

There's been a "hardware index" in watchpoints which we initialize
if the remote gdb stub happens to give us the hardware index of the
watchpoint register, but there way lldb receives this index depended
on non-standard additions to the watchpoint packet.  Instead, I
have WatchpointResources contain an ID which is the hardware target
register used for that watchpoint, which I initialize heuristically
assuming that the stub will set new watchpoints in the first available
slot starting at 0, and when a watchpoint is deleted/disabled, that
spot will now be unused.  It's only a heuristic, but it's no worse
than our current sometimes-we-can-get-an-index system.

With this update to the PR, this patch is still not breaking a
user's watchpoint up into multiple WatchpointResources, it is not
sharing a single WatchpointResource for multiple Watchpoints, and
it is still not updating all Watchpoints for a Resource when an
access is received.

I am starting to be a little worried about the work needing to be
done in the Process plugin to finish this feature - I will have
no ability to build or test ProcessWindows or NativeProcessWindows.
I will do the rest of the work trying to keep it in base classes
to reduce the amount of change needed in those, but I doubt it will
be enough that I can change those blindly.

jasonmolenda commented:
@bulbazord @JDevlieghere OK this is the second set of refinement where it's closer to the structure of BreakpointSites.  It's still not actually doing anything -- each one WatchpointResource corresponds to one Watchpoint, a single Watchpoint's request is not broken into multiple WatchpointResources, and when a watchpoint access is hit, I'm not iterating all of the Watchpoints (lol, all 1 of them) to bump their hit count, run conditions, etc.  These remaining pieces of work will be smaller scale, more complicated I think, and it will be easier to consider those parts if the patches are small.

jasonmolenda commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+  // ordering.  If a Process can correctly identify the hardware watchpoint
+  // register index when it has created the Resource, it may initialize it
+  // before it is inserted in the WatchpointResourceList.
+  void SetID(lldb::wp_resource_id_t);
```
When the Process method is given a watchpoint request from the user, which must be broken into one or more WatchpointResources (or an existing WatchpointResource reused), we can only get the hardware register number by looking at the IDs of currently allocated WatchpointResources in the WatchpointResourceList, which start at 0.  I assume that the remote stub will allocate its watchpoints in the hardware registers starting at index 0, and when removing/disabling them, will leave that slot unused until another request comes along, at which point the stub will start from 0 and look for the first unused hardware register.

In the specific case of gdb remote serial protocol (ProcessGDBRemote), it has none of this.  The Z2/Z3/Z4 packets don't tell you what hardware register was used.  

Today, lldb sets the hardware watchpoint index based on the watchpoint access stop.  SOME ways of passing the watchpoint trap back to lldb can communicate the hardware watchpoint index.  The standard gdb remote serial protocol does not with its `watch/rwatch/awatch:address` key-value in the stop packet.  lldb-server has a different way of reporting watchpoints, `reason:watchpoint` and `description:addr[ index]` where the address is sent in the description: and it may be followed by a space and then the index number of the hardware watchpoint that was triggered.  For macOS systems, the mach exception data is sent at the stop, and the hardware index is encoded as a piece of data in that.

In reality, in both cases I believe, the stubs are actually getting an *address* of the access, and trying to figure out which one of the hardware watchpoints they have set was the actual one triggered.  Aarch64 will some day have a way reporting the actual hardware index triggered instead of the exception address (it may not even give an exception address at all), but today it's all a fiction.

So basically, today, you don't really know the hardware watchpoint index until it's been hit.  And even then, it's being deduced by the remote stub, which lldb could do just as well.  The idea to expose this piece of data we don't actually have was not a good one, IMO, but it's in SB API so I'm not tackling that for now.

If I wanted to have the Process plugin look at the IDs currently used in its WatchpointResourceList, use the available numbers for the WatchpointResources it is creating, and if they are set successfully in the inferior, add them to the WatchpointResourceList, I would need to lock the WatchpointResourceList for the duration of those operations, and I wasn't sure that was a good idea.

(sorry for the long reply, I spent so much time trying to decide what to do with these fictitious hardware index numbers we present in lldb)

jasonmolenda replied:
To be honest, I'm VERY close to just removing the hardware index numbers altogether and having SBWatchpoint::GetHardwareIndex return UINT32_MAX.  It might be the better idea than these convoluted things I'm doing to try to heuristically guess the hw index.

medismailben replied:
Thanks for the explanation, that makes more sense to me now.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+
+  ~WatchpointResource();
+
+  void GetMemoryRange(lldb::addr_t &addr, size_t &size) const;
```
Instead of out parameters, you could do `std::pair<addr_t, size_t>`.
But why have this and `GetAddress()` + `GetByteSize()`?

jasonmolenda replied:
I originally wrote the WatchpointResource::GetMemoryRange that returned both via out parameters, then at a later point I added a GetAddress & GetByteSize methods meaning to remove GetMemoryRange which I wasn't happy with... but forgot.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+
+  size_t GetByteSize() const;
+
+  void GetType(bool &read, bool &write) const;
```
Suggestion: Return `std::pair<bool, bool>` instead of passing out parameters.

jasonmolenda replied:
idk about this one, `GetType(bool &read, bool &write)` makes the order easy to see, and matches the `SetType (bool read, bool write)` whereas the pair would be less explicitly named.  I should follow the Watchpoint class method and have `bool WatchpointResourceRead()`, `bool WatchpointResourceWrite()`.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+  ///
+  /// \return
+  ///    The number of owners.
+  size_t GetNumberOfOwners();
```
This can be `const`-qualified right?

jasonmolenda replied:
I lock the object's mutex in all of these methods.

bulbazord replied:
Ah I could have sworn the mutex was marked as `mutable`. How are things like `GetType` and `GetByteSize` marked `const` then? Do they not lock?

bulbazord replied:
Oh wait, the mutex is to protect the Owners. Nevermind!

jasonmolenda replied:
tbh there could be an argument for locking the whole object because a WatchpointResource may be expanded to service multiple watchpoints (one watchpoint watches bytes 0-1, a second watchpoint watches bytes 2-3) and then reduced again when one of those watchpoints is deleted/disabled.

jasonmolenda replied:
none of that is done yet though.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+  ///    The number of owners.
+  size_t GetNumberOfOwners();
+
+  /// This method returns the Watchpoint at index \a index using this
+  /// Resource.  The owners are listed ordinally from 0 to
+  /// GetNumberOfOwners() - 1 so you can use this method to iterate over the
+  /// owners.
+  ///
+  /// \param[in] idx
+  ///     The index in the list of owners for which you wish the owner location.
+  ///
+  /// \return
+  ///    The Watchpoint at that index.
+  lldb::WatchpointSP GetOwnerAtIndex(size_t idx);
```
Instead of exposing this primitive, maybe it would make more sense to return an iterator or an iterator range?
Something like:
```
IteratorRange<lldb::WatchpointSP> Owners();
```
IMO this would make it easier to write bug-free code since we're not going to have to manually set up index iteration logic.

jasonmolenda replied:
Just getting back to the review comments.  I remember writing something about this last week but we have some API to select specific BreakpointLocations that own the BreakpointSite and the code to do that is most naturally exposed as "GetAtIndex" style access.  I'm trying to create the WatchpointResource class similar to BreakpointSite so I can follow the same coding style for these Watchpoints that own the WatchpointResources.  I may end up being able to drop the GetAtIndex method later if it's natural go access it via an Owners method, but I'm starting with the assumption that this will be more natural.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+  ///     true if this resource's owners includes the watchpoint.
+  bool OwnersContains(const lldb_private::Watchpoint *wp);
+
+  /// This method copies the watchpoint resource's owners into a new collection.
+  /// It does this while the owners mutex is locked.
+  ///
+  /// \param[out] out_collection
+  ///    The BreakpointLocationCollection into which to put the owners
+  ///    of this breakpoint site.
+  ///
+  /// \return
+  ///    The number of elements copied into out_collection.
+  size_t CopyOwnersList(WatchpointCollection &out_collection);
```
Why return the size instead of returning just a new `WatchpointCollection` object? That way callers don't have to create a new one and pass it in, they can just grab it from this method directly.

jasonmolenda replied:
Hm, I do not have a lot of confidence about avoiding an extra copy of the vector if I return by value.

bulbazord replied:
It should take advantage of copy elision. Specifically we should be able to take advantage of NRVO.

https://en.cppreference.com/w/cpp/language/copy_elision

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+  lldb::addr_t m_addr;
+  size_t m_size;
+
+  bool m_watch_read;  // true if we stop when the watched data is read from
+  bool m_watch_write; // true if we stop when the watched data is written to
```
IMO these comments are unnecessary, the names of the variables already describe the intended semantics quite well I think.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,140 @@
+  bool m_watch_read;  // true if we stop when the watched data is read from
+  bool m_watch_write; // true if we stop when the watched data is written to
+
+  // The watchpoints using this WatchpointResource.
+  WatchpointCollection m_owners;
```
Instead of calling them "Owners", maybe the name "Users" makes more sense? I guess you could say that the watchpoints own this Resource, but in this comment you're explaining that owners are watchpoints that *use* this resource. WDYT?

jasonmolenda replied:
I was never thrilled with the Owners nomenclature but I was trying to follow the design of BreakpointSites which are owned by one or more BreakpointLocations. My preferences snuck in to the comment here.

bulbazord commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -266,33 +268,73 @@ void Watchpoint::Dump(Stream *s) const {
     const char *old_value_cstr = m_old_value_sp->GetValueAsCString();
-    if (old_value_cstr && old_value_cstr[0])
-      s->Printf("\n%sold value: %s", prefix, old_value_cstr);
-    else {
+    if (old_value_cstr) {
```
You can fold this into one line:
```
if (auto *old_value_cstr = m_old_value_sp->GetValueAsCString())
```
(use of auto here is personal preference, the idea is independent of that)

bulbazord commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -266,33 +268,73 @@ void Watchpoint::Dump(Stream *s) const {
+    } else {
       const char *old_summary_cstr = m_old_value_sp->GetSummaryAsCString();
-      if (old_summary_cstr && old_summary_cstr[0])
-        s->Printf("\n%sold value: %s", prefix, old_summary_cstr);
+      if (old_summary_cstr)
```
Same here

bulbazord commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -266,33 +268,73 @@ void Watchpoint::Dump(Stream *s) const {
+
     const char *new_value_cstr = m_new_value_sp->GetValueAsCString();
-    if (new_value_cstr && new_value_cstr[0])
-      s->Printf("\n%snew value: %s", prefix, new_value_cstr);
+    if (new_value_cstr)
+      values_ss.Printf("new value: %s", new_value_cstr);
```
Same here

bulbazord commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -266,33 +268,73 @@ void Watchpoint::Dump(Stream *s) const {
       const char *new_summary_cstr = m_new_value_sp->GetSummaryAsCString();
-      if (new_summary_cstr && new_summary_cstr[0])
-        s->Printf("\n%snew value: %s", prefix, new_summary_cstr);
+      if (new_summary_cstr)
+        values_ss.Printf("new value: %s", new_summary_cstr);
```
Same here

bulbazord commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -331,6 +373,16 @@ void Watchpoint::DumpWithLevel(Stream *s,
+        ->GetWatchpointResourceList()
+        .FindByWatchpoint(this)
+        ->GetID();
+  else
+    return UINT32_MAX;
```
no need for the else since the if block returns something.

bulbazord commented on lldb/source/Breakpoint/WatchpointCollection.cpp:
```diff
@@ -0,0 +1,90 @@
+WatchpointCollection &
+WatchpointCollection::operator=(const WatchpointCollection &rhs) {
+  if (this != &rhs) {
+    std::lock(m_collection_mutex, rhs.m_collection_mutex);
+    std::lock_guard<std::mutex> lhs_guard(m_collection_mutex, std::adopt_lock);
+    std::lock_guard<std::mutex> rhs_guard(rhs.m_collection_mutex,
+                                          std::adopt_lock);
```
You can use `std::scoped_lock` for this now that we're using C++17! 😄

bulbazord commented on lldb/source/Breakpoint/WatchpointCollection.cpp:
```diff
@@ -0,0 +1,90 @@
+
+bool WatchpointCollection::Contains(const WatchpointSP &wp_sp) {
+  std::lock_guard<std::mutex> guard(m_collection_mutex);
+  const size_t size = m_collection.size();
+  for (size_t i = 0; i < size; ++i) {
+    if (m_collection[i] == wp_sp)
+      return true;
+  }
```
Why not use iterators here? You can also use `llvm::is_contained` too.

bulbazord commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -0,0 +1,96 @@
+void WatchpointResource::SetID(wp_resource_id_t id) { m_id = id; }
+
+bool WatchpointResource::Contains(addr_t addr) {
+  if (addr >= m_addr && addr < m_addr + m_size)
+    return true;
+  return false;
```
`return add >= m_addr && add < m_addr + m_size;`

jasonmolenda replied:
yeah this kind of thing always makes me think 10% harder when I'm reading code, but I know I'm in the minority.

bulbazord commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -0,0 +1,96 @@
+WatchpointResource::CopyOwnersList(WatchpointCollection &out_collection) {
+  std::lock_guard<std::recursive_mutex> guard(m_owners_mutex);
+  const size_t size = m_owners.GetSize();
+  for (size_t i = 0; i < size; ++i) {
+    out_collection.Add(m_owners.GetByIndex(i));
+  }
```
1. no need for braces per llvm style guidelines.
2. Instead of iterating by index, you could do something like:
```
for (auto watchpoint_sp : m_owners)
  out_collection.Add(watchpoint_sp);
```
or something to this effect (use of auto is personal preference)

bulbazord commented on lldb/source/Plugins/Process/Utility/StopInfoMachException.cpp:
```diff
@@ -491,14 +491,13 @@ static StopInfoSP GetStopInfoForHardwareBP(Thread &thread, Target *target,
                                            uint64_t exc_sub_sub_code) {
   // Try hardware watchpoint.
   if (target) {
+    // LWP_TODO: We need to find the WatchpointResource that matches
```
`LWP_TODO` doesn't mean much to a lot of contributors, maybe you can change this to something like `TODO(Large Watchpoint Support)`? Mostly to avoid acronyms that aren't written down anywhere else.

jasonmolenda replied:
Yeah good point my idea was that these are temporary placeholders as I roll this out, so they wouldn't be in the source for very long. (this feature is not really useful with all the parts landed so it's unlikely I'll be distracted by some other issue ....)

bulbazord commented on lldb/source/Plugins/Process/Windows/Common/ProcessWindows.cpp:
```diff
@@ -839,11 +841,11 @@ std::optional<uint32_t> ProcessWindows::GetWatchpointSlotCount() {
 
-  if (wp->IsEnabled()) {
-    wp->SetEnabled(true, notify);
+  if (wp_sp->IsEnabled()) {
```
Should we check if `wp_sp` is even valid first?

jasonmolenda replied:
The original code didn't check it, but more importantly on this topic, I have real concern about touching any of ProcessWindows, NativeProcessWindows because I can't build or test it directly myself.  I have a feeling I'll keep most/all of the existing interfaces for creating a StopInfo etc so the Windows code can continue to work unmodified.

bulbazord commented on lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:
```diff
@@ -3105,102 +3122,184 @@ Status ProcessGDBRemote::DisableBreakpointSite(BreakpointSite *bp_site) {
+  bool read, write;
+  wp_res_sp->GetType(read, write);
+
+  assert((read || write) && "read and write cannot both be false.");
```
The assertion message here could be fleshed out a bit more. Something like `"WatchpointResource type is neither read nor write"`.

bulbazord commented on lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:
```diff
@@ -3105,102 +3122,184 @@ Status ProcessGDBRemote::DisableBreakpointSite(BreakpointSite *bp_site) {
-  if (wp) {
-    user_id_t watchID = wp->GetID();
-    addr_t addr = wp->GetLoadAddress();
+  if (wp_sp) {
```
Since you're already refactoring this part, maybe you could flip the condition and do the error handling up front. It's small, but it saves a level of indentation in an already large code block.

bulbazord reviewed (commented):
A lot of the `WatchpointResourceList::FindBy` methods can probably be rewritten with `std::find_if` (or `llvm::find_if` which is a little easier to use IMO).

Overall, looks like a pretty nice mostly-NFC refactor! 😄

jasonmolenda committed:
Address Alex's feedback

Remove WatchpointCollection, this was little more than
a std::vector, so now it's a typedef to that.  In the
Breakpoint case where a BreakpointSite has a
BreakpointLocationCollection, the lldb container class
was needed.  Clean up WatchpointResource class API.
A few smaller fixups like using early returns to avoid
unnecessary indentation.

bulbazord commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -53,44 +49,52 @@ bool WatchpointResource::Contains(addr_t addr) {
   std::lock_guard<std::recursive_mutex> guard(m_owners_mutex);
-  return m_owners.Contains(wp_sp);
+  const auto &it = std::find(m_owners.begin(), m_owners.end(), wp_sp);
+  if (it != m_owners.end())
+    return true;
+  return false;
```
`return it != m_owners.end();`

bulbazord commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -53,44 +49,52 @@ bool WatchpointResource::Contains(addr_t addr) {
 bool WatchpointResource::OwnersContains(const Watchpoint *wp) {
   std::lock_guard<std::recursive_mutex> guard(m_owners_mutex);
-  return m_owners.Contains(wp);
+  for (WatchpointCollection::const_iterator it = m_owners.begin();
+       it != m_owners.end(); ++it)
+    if ((*it).get() == wp)
+      return true;
+  return false;
```
`WatchpointCollection` got removed right? I think you'll need to rewrite it... I'd suggest using `find_if`

bulbazord commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -53,44 +49,52 @@ bool WatchpointResource::Contains(addr_t addr) {
     return {};
 
-  return m_owners.GetByIndex(idx);
+  return m_owners[idx];
```
Looking over this again, is there a use case where you would want to reference a specific owner by index instead of using `Owners()`?

jasonmolenda replied:
Right now nothing is using the WatchpointResource owners - I need to add that in a follow on patch when a watchpoint has been triggered.  BreakpointSite uses this method in many places, but I agree that the Owners() method could be used.  We lose the range check/possible-assert in GetOwnerAtIndex() but hopefully that shouldn't actually happen...

jasonmolenda replied:
I'm not completely clear what having WatchpointResource::Owners() returning a WatchpointIterable is going to do beyond allowing someone to write `for (WatchpointSP &wp_sp : wp_resource_sp->Owners())` - it's not exposing the underlying std::vector so I can't call `wp_resource_sp->Owners().size()` or `wp_resource_sp->Owners()[2]` is it?  Right now the WatchpointResource has methods like `AddOwner`, `RemoveOwner`, `GetNumberOfOwners`, `GetOwnerAtIndex`, and some methods to check if the list of owners contains a watchpoint; should callers be able to lock the Resource's mutex and get a reference to the std::vector and query/manipulate it directly?

bulbazord replied:
Well, I'm not entirely sure how we're going to want to access the Owners list, so I'd like to understand more how you plan on using it. If you want to do something like `wp_resource_sp->Owners()[2]` then this method makes sense to keep around, but I'd also be curious to know why you would want to access the 3rd owner directly. What I would like to avoid is writing a loop like this:
```
for (int i = 0; i < wp_resource_sp.GetNumberOfOwners(); i++) {
  WatchpointSP wp_sp = wp_resource_sp->GetOwnerAtIndex(i);
  // Use wp_sp
}
```
The two main reasons I am not a fan of this pattern:
1. As you pointed out, one would need to grab the `wp_resource_sp` owner mutex to do this safely since another thread may call `RemoveOwner` in the middle of this loop. I can't say I'm a fan of exposing the vector primarily because it would mean we'd have to expose the mutex too.
2. Off-by-one errors are fairly easy to put in by mistake. Doubly so if we ever iterate in reverse (i >0 vs i >=0 for the termination condition).

If we don't need to be able to access any of the owners at an arbitrary index, I would suggest removing it so nobody can write that kind of loop. WDYT?

jasonmolenda replied:
This is a good point.  I am mostly trying to copy the API of BreakpointSite which has methods like this, and that's how its Owners are accessed today in different parts of lldb.  I haven't gotten to the point where I use the Owners anywhere yet, so I'm not locked in to the same API that BreakpointSite exposes.  We have things like `SBThread::GetStopReasonDataAtIndex()` which for Breakpoints exposes all of the Breakpoints that were owning the BreakpointSite that was hit.  We have other places like `StackFrameList::ResetCurrentInlinedDepth` which is iterating over a BreakpointSite's owners to see if any of them are internal; the IterableWatchpoint returned by `WatchpointResources::Owners` would be a cleaner approach in this case.  We have some places in the code that need to know the address of the breakpoint that was hit and are grabbing the first BreakpointLocation associated with the BreakpointSite, e.g. `PlatformDarwin::GetSoftwareBreakpointTrapOpcode` (for some arm/thumb reason).  In `ThreadPlanCallFunction::DoPlanExplainsStop` we're iterating over the Owners to log about them and detect if any are Internal breakpoint locations, which could be done with an Iterable.

jasonmolenda committed:
Update WatchpointResource OwnersContains methods

Suggestions from Alex.

bulbazord commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -67,18 +67,15 @@ size_t WatchpointResource::GetNumberOfOwners() {
-    if ((*it).get() == wp)
-      return true;
-  return false;
+  WatchpointCollection::const_iterator match =
```
Looks fine, but the type should be different right? I think you removed `WatchpointCollection` in the 3rd commit

jasonmolenda replied:
It's a typedef to std::vector<WatchpointSP>.

bulbazord replied:
My bad... 🤦

JDevlieghere commented on lldb/source/Breakpoint/WatchpointResourceList.cpp:
```diff
@@ -0,0 +1,154 @@
+  for (collection::iterator pos = m_resources.begin(); pos != m_resources.end();
+       ++pos)
+    if ((*pos)->Contains(addr))
+      return *pos;
```
Can you use a range-based for loop here and below?

jasonmolenda commented on lldb/include/lldb/Breakpoint/WatchpointResource.h:
```diff
@@ -0,0 +1,155 @@
+  WatchpointCollection m_owners;
+
+  std::recursive_mutex
+      m_owners_mutex; ///< This mutex protects the owners collection.
```
No, I was just copying BreakpointSite when I used that.

jasonmolenda commented on lldb/source/Breakpoint/Watchpoint.cpp:
```diff
@@ -266,33 +268,73 @@ void Watchpoint::Dump(Stream *s) const {
-    prefix = "";
-  }
+bool Watchpoint::DumpSnapshots(Stream *s, const char *prefix) const {
+  bool printed_anything = false;
```
The convention for these methods seems to be to return a stream without a final end-of-line character.  In StopInfo.cpp we call this method and check if it printed anything, and if so, print an EOL.  In Watchpoint::DumpWithLevel we ignore the return type from DumpSnapshots, but that's only because every line it adds is assumed to end without an EOL.  I don't really need a variable in this method, I could return false in the two places it is false and true in the one place it is true.

jasonmolenda commented on lldb/include/lldb/Breakpoint/WatchpointResourceList.h:
```diff
@@ -0,0 +1,146 @@
+//===-- WatchpointResourceList.h --------------------------------*- C++ -*-===//
```
I think it's the same except for the obvious BreakpointSite/WatchpointResource and BreakpointLocation/Watchpoint.  It's a little tricky reading through BreakpointSiteList.h to tell if all of the arguments refer to BreakpointLocations (they do it by id values) or if they might refer to Breakpoints - they seem to all use the same `break_id_t`?  (and `break_id_t` is also used for the BreakpointSite numbers); sometimes the arguments are called `bp_id` and sometimes they're called `BreakID` but I THINK they're always referring to BreakpointLocations.

It'd be nice if BreakpointSite and WatchpointResource could inherit from the same base class and have the StopPointList work in terms of the base class methods, but I think there's enough difference between these two that this might not work very cleanly.

Maybe a templated StopPointList could work though.  I'll need to look at that.

jasonmolenda committed:
Address this second round of comments from Alex and Jonas

Jonas' suggestion of trying to unify
BreakpointSiteList/WatchpointResourceList is something I still
need to look at.

jasonmolenda committed:
Use foreach iterators in WatchpointResourceList

Missed this review comment by Jonas.

jasonmolenda committed:
Address Jonas' 1st review comments, biggest: StopPointSiteList template

My original patch had a BreakpointSiteList and a WatchpointResourceList
which are pretty simple collections with some unique lookup methods
for finding things by the watchpoint resource/breakpoint site IDs.
Jonas asked that to be a template specialized for both classes;
this patch adds a StopPointSiteList template class.

The BreakpointSite / WatchpointResources have "owners".  A
BreakpointSite is owned by the BreakpointLocations that it is
used for.  A WatchpointResource is owned by the Watchpoints
at that memory range.

For some reason the owner name doesn't seem clear to me, and I
thought of a politican who has constituents that they serve.  I
tried out a rename from a Site having "owners" to having "constituents".
I don't know if it's really better in any way.  And the "owners"
terminology is used in other places that Jim wrote; BreakpointLocations
have owners (Breakpoints).  ThreadPlans have owners (Threads).
I don't feel strongly about this and "constituent" is easy to global
search and replace back if anyone thinks this isn't a good choice.

JDevlieghere commented on lldb/source/Breakpoint/StopPointSiteList.cpp:
```diff
@@ -0,0 +1,235 @@
+  typename collection::iterator iter = m_site_list.find(site_load_addr);
+
+  if (iter == m_site_list.end()) {
+#if 0
+    m_site_list.insert(iter, collection::value_type(site_load_addr, site));
+#endif
```
Remove.

JDevlieghere commented on lldb/source/Breakpoint/StopPointSiteList.cpp:
```diff
@@ -0,0 +1,235 @@
+template <typename StopPointSite>
+bool StopPointSiteList<StopPointSite>::ShouldStop(
+    StoppointCallbackContext *context, typename StopPointSite::SiteID site_id) {
+  StopPointSiteSP site_sp(FindByID(site_id));
+  if (site_sp) {
```
```if(StopPointSiteSP site_sp = FindByID(site_id))```

JDevlieghere commented on lldb/source/Breakpoint/StopPointSiteList.cpp:
```diff
@@ -0,0 +1,235 @@
+template <typename StopPointSite>
+void StopPointSiteList<StopPointSite>::Dump(Stream *s) const {
+  s->Printf("%p: ", static_cast<const void *>(this));
+  // s->Indent();
```
Remove

JDevlieghere commented on lldb/source/Breakpoint/StopPointSiteList.cpp:
```diff
@@ -0,0 +1,235 @@
+
+  upper = m_site_list.upper_bound(upper_bound);
+
+  for (pos = lower; pos != upper; pos++) {
+    site_list.Add((*pos).second);
+  }
```
No braces

JDevlieghere commented on lldb/source/Breakpoint/StopPointSiteList.cpp:
```diff
@@ -0,0 +1,235 @@
+  return sites;
+}
+
+namespace lldb_private {
+template class StopPointSiteList<BreakpointSite>;
+template class StopPointSiteList<WatchpointResource>;
+} // namespace lldb_private
```
Although this is perfectly fine way to allow you to define your templated methods in the `.cpp` file, as far as I can tell we don't do this anywhere else in LLDB. I found some prior art in LLVM and the fact that you have exactly two instantiations helps but given that this (1) doesn't pull in all that many dependencies and (2) probably ins't included soo much that we have to worry about compile time, I would've just implemented this in the header.

JDevlieghere commented on lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:
```diff
@@ -1787,30 +1788,48 @@ ThreadSP ProcessGDBRemote::SetThreadStopInfo(
+          if (!wp_resource_sp) {
             Log *log(GetLog(GDBRLog::Watchpoints));
             LLDB_LOGF(log, "failed to find watchpoint");
+            abort(); // LWP_TODO FIXME don't continue executing this block if
```
??

jasonmolenda replied:
My refactored code added a new nullptr deref, when I saw what I'd done I threw in an abort and a FIXME comment to fix that.  But I see how to fix it, let me do that.

JDevlieghere reviewed (approved):
LGTM with a few nits. I like "constituents". 

This is a pretty big patch which makes reviewing it challenging. I know it's a big change that touches a lot of things but I'm sure that this could've been broken up into smaller patches if you keep that goal in mind from the beginning. Something to look out for in the future.

jasonmolenda committed:
Address Jonas' second round of comments.

jasonmolenda commented:
> LGTM with a few nits. I like "constituents".
> 
> This is a pretty big patch which makes reviewing it challenging. I know it's a big change that touches a lot of things but I'm sure that this could've been broken up into smaller patches if you keep that goal in mind from the beginning. Something to look out for in the future.

Thanks for the feedback.  Yeah originally this patch was a bit smaller but it has Grown as I've addressed (correct, good) feedback from everyone and now it's a little bit of a monster.  I'm surely going to have to rebase it before I can merge.

bulbazord reviewed (approved):
LGTM as well. I like the name Constituents quite a bit as well! 😄

I think most feedback at this point can probably be handled in follow-ups, esp. since some of it pertains to existing interfaces.

jasonmolenda committed:
Remove orphaned wp hardware index tracking

jasonmolenda committed:
Move the StopPointSiteList template methods into hdr

I have one method which is only defined for BreakpointSites
which I specialize in the .cpp file still, everything else is
in the .h header file.

I still have one test failure in TestWatchpointCount.py that
should be easy to fix, then I might be done with this PR.

jasonmolenda committed:
Fix WatchpointResource ID setting

I still had my original attempts to intuit the most likely hardware
index number when adding a WatchpointResource to the process resource
list.  I've dropped that whole scheme in a separate commit; this
commit correctly goes back to allocating the IDs sequentially for
the Resources.

sylvestre commented:
Seems that it caused https://github.com/llvm/llvm-project/issues/73580 no?

dyung commented:
> Seems that it caused #73580 no?

I believe so. @jasonmolenda I see you attempted to make a fix in a0a1ff3ab40e347589b4e27d8fd350c600526735, however I still have two buildbots that are failing to build. Can you either fix it or revert if you need time to investigate?

https://lab.llvm.org/buildbot/#/builders/217/builds/31915
https://lab.llvm.org/buildbot/#/builders/243/builds/16118
